building_sim_plugins/building_ignition_plugins/src/crowd_simulator.cpp

```diff
@@ -384,33 +398,65 @@ void CrowdSimulatorPlugin::_update_internal_object(
+  auto anim_name_comp =
+    ecm.Component<ignition::gazebo::components::AnimationName>(entity);
+  if (nullptr == anim_name_comp)
+  {
+    RCLCPP_ERROR(_crowd_sim_interface->logger(),
+      "Model [" + obj_ptr->model_name + "] has no AnimationName component.");
+    exit(EXIT_FAILURE);
+  }
+  auto anim_time_comp =
+    ecm.Component<ignition::gazebo::components::AnimationTime>(entity);
+  if (nullptr == anim_name_comp)
+  {
+    RCLCPP_ERROR(_crowd_sim_interface->logger(),
+      "Model [" + obj_ptr->model_name + "] has no AnimationTime component.");
+    exit(EXIT_FAILURE);
+  }
+  auto actor_comp =
+    ecm.Component<ignition::gazebo::components::Actor>(entity);
 
+  ignition::math::Pose3d current_pose = traj_pose_comp->Data();
   auto distance_traveled_vector = agent_pose.Pos() - current_pose.Pos();
   // might need future work on 3D case
   // the center of human has a z_elevation, which will make the human keep walking even if he reached the target
   distance_traveled_vector.Z(0.0);
   double distance_traveled = distance_traveled_vector.Length();
 
+  // switch animation
+  auto idle_animation =
+    _crowd_sim_interface->_model_type_db_ptr->get(obj_ptr->type_name)->
+    idle_animation;
+  if (distance_traveled - _crowd_sim_interface->get_switch_anim_distance_th() <
+    1e-6 &&
+    !idle_animation.empty() )
+  {
+    anim_name_comp->Data() = idle_animation;
+    anim_time_comp->Data() +=
+      std::chrono::duration_cast<std::chrono::steady_clock::duration>(
+      std::chrono::duration<double>(delta_sim_time));
+    // lock yaw angle
+    agent_pose.Rot() = current_pose.Rot();
+  }
+  else
+  {
+    anim_name_comp->Data() =
+      _crowd_sim_interface->_model_type_db_ptr->get(obj_ptr->type_name)
+      ->animation;
+    anim_time_comp->Data() +=
+      std::chrono::duration_cast<std::chrono::steady_clock::duration>(
+      std::chrono::duration<double>(distance_traveled / animation_speed));
+  }
+
   // set trajectory
-  *traj_pose_comp = ignition::gazebo::components::TrajectoryPose(agent_pose);
+  traj_pose_comp->Data() = agent_pose;
   ecm.SetChanged(entity,
     ignition::gazebo::components::TrajectoryPose::typeId,
     ignition::gazebo::ComponentState::OneTimeChange);
-  //set animation
-  auto anim_time_comp =
-    ecm.Component<ignition::gazebo::components::AnimationTime>(entity);
-  if (nullptr == anim_time_comp)
-  {
-    RCLCPP_ERROR(_crowd_sim_interface->logger(),
-      "Model [" + obj_ptr->model_name + "] has no AnimationTime component");
-    exit(EXIT_FAILURE);
-  }
-  auto anim_time = anim_time_comp->Data() +
-    std::chrono::duration_cast<std::chrono::steady_clock::duration>(
-    std::chrono::duration<double>(distance_traveled / animation_speed));
-
-  *anim_time_comp = ignition::gazebo::components::AnimationTime(anim_time);
+  ecm.SetChanged(entity,
+    ignition::gazebo::components::AnimationName::typeId,
+    ignition::gazebo::ComponentState::OneTimeChange);
```

Comment:
Moving into a state machine should make it possible to call this instruction only once on state transition instead of every iteration.